tiledb/sm/query/ast/query_ast.h

```diff
@@ -0,0 +1,275 @@
+      condition_value_view_ = UntypedDatumView(
+          (rhs.condition_value_view_.content() == nullptr ?
+               nullptr :
+               condition_value_data_.data()),
+          condition_value_data_.size());
+      op_ = rhs.op_;
+    }
+    return *this;
+  }
+
+  /** Move-assignment operator. */
+  ASTNodeVal& operator=(ASTNodeVal&& rhs) {
+    field_name_ = std::move(rhs.field_name_);
+    condition_value_data_ = std::move(rhs.condition_value_data_);
+    condition_value_view_ = std::move(rhs.condition_value_view_);
+    op_ = rhs.op_;
+
+    return *this;
+  }
+
+  /** Returns the tag attached to the node. */
+  ASTNodeTag get_tag() {
+    return tag_;
+  }
+
+  /** Returns a deep copy of the node. */
+  tdb_unique_ptr<ASTNode> clone() {
+    return tdb_unique_ptr<ASTNode>(tdb_new(
+        ASTNodeVal,
+        field_name_,
+        condition_value_data_.data(),
+        condition_value_data_.size(),
+        op_));
+  }
+
+  /** The attribute name. */
+  std::string field_name_;
+
+  /** The value data. */
+  ByteVecValue condition_value_data_;
+
+  /** A view of the value data. */
+  UntypedDatumView condition_value_view_;
+
+  /** The comparison operator. */
+  QueryConditionOp op_;
+
+ private:
+  static const ASTNodeTag tag_{ASTNodeTag::VAL};
+};
+
+class ASTNodeExpr : public ASTNode {
+ public:
+  ASTNodeExpr(
+      std::vector<tdb_unique_ptr<ASTNode>>&& nodes,
+      QueryConditionCombinationOp c_op)
+      : combination_op_(c_op) {
+    for (const auto& elem : nodes) {
+      nodes_.push_back(elem->clone());
+    }
+  }
+
+  /** Default destructor. */
+  ~ASTNodeExpr() {
+  }
+
+  /** Copy constructor. */
+  ASTNodeExpr(const ASTNodeExpr& rhs)
+      : combination_op_(rhs.combination_op_) {
+    for (const auto& elem : rhs.nodes_) {
+      nodes_.push_back(elem->clone());
+    }
+  };
+
+  /** Move constructor. */
+  ASTNodeExpr(ASTNodeExpr&& rhs)
+      : nodes_(std::move(rhs.nodes_))
+      , combination_op_(rhs.combination_op_){};
+
+  /** Assignment operator. */
+  ASTNodeExpr& operator=(const ASTNodeExpr& rhs) {
+    if (this != &rhs) {
+      for (const auto& elem : rhs.nodes_) {
+        nodes_.push_back(elem->clone());
+      }
+      combination_op_ = rhs.combination_op_;
+    }
+    return *this;
+  }
+
+  /** Move-assignment operator. */
+  ASTNodeExpr& operator=(ASTNodeExpr&& rhs) {
+    nodes_ = std::move(rhs.nodes_);
+    combination_op_ = rhs.combination_op_;
+    return *this;
+  }
+
+  /** Returns the tag attached to the node. */
+  ASTNodeTag get_tag() {
+    return tag_;
+  }
+
+  /** Returns a deep copy of the node. */
+  tdb_unique_ptr<ASTNode> clone() {
+    std::vector<tdb_unique_ptr<ASTNode>> nodes_copy;
+    for (const auto& node : nodes_) {
+      nodes_copy.push_back(node->clone());
+    }
+    return tdb_unique_ptr<ASTNode>(
+        tdb_new(ASTNodeExpr, std::move(nodes_copy), combination_op_));
+  }
+
+  /** The node list **/
+  std::vector<tdb_unique_ptr<ASTNode>> nodes_;
```

Comment:
For my understanding, is this a vector of `ASTNode` or a vector of `ASTNodeVal`? In other words is every entry in the vector a Value node or it can be an Expression as well?

Reply:
It can also be an expression node.